design/api/build-spi/swagger.yaml

```diff
@@ -0,0 +1,431 @@
+swagger: '2.0'
+
+info:
+  version: "0.0.0"
+  title: Build SPI
+  description: |
+    Build SPI specification.
+
+securityDefinitions:
+  TargetUrl:
+    type: "apiKey"
+    name: "targetUrl"
+    in: header
+  TargetToken:
+    type: "apiKey"
+    name: "targetToken"
+    in: header
+
+security:
+- TargetUrl: []
+- TargetToken: []
+
+# TODO:
+# - support rate limiting in reponses for public providers to avoid DDoS
+
+paths:
+  /namespace/{namespace}/buildconfigs:
+    parameters:
+    - $ref: "#/parameters/Namespace"
+    post:
+      summary: "Create a BuildConfig."
+      description: "Create a BuildConfig."
+      parameters:
+      - $ref: "#/parameters/BuildConfig"
+      responses:
+        204:
+          $ref: "#/responses/204"
+        401:
+          $ref: "#/responses/401"
+        500:
+          $ref: "#/responses/500"
+  /namespace/{namespace}/buildconfigs/{name}:
+    parameters:
+    - $ref: "#/parameters/Namespace"
+    - $ref: "#/parameters/Name"
+    get:
+      summary: "Read a BuildConfig."
+      description: "Read a BuildConfig."
+      responses:
+        200:
+          description: "OK"
+          schema:
+            $ref: "#/definitions/BuildConfig"
+        401:
+          $ref: "#/responses/401"
+        500:
+          $ref: "#/responses/500"
+    delete:
+      summary: "Delete a BuildConfig."
+      description: "Delete a BuildConfig."
+      responses:
+        204:
+          $ref: "#/responses/204"
+        401:
+          $ref: "#/responses/401"
+        500:
+          $ref: "#/responses/500"
+    put:
+      summary: "Replace the specified BuildConfig"
+      description: "Replace the specified BuildConfig"
+      parameters:
+      - $ref: "#/parameters/BuildConfig"
+      responses:
+        204:
+          $ref: "#/responses/204"
+        401:
+          $ref: "#/responses/401"
+        500:
+          $ref: "#/responses/500"
+  /namespace/{namespace}/buildconfigs/{name}/instantiate:
+    parameters:
+    - $ref: "#/parameters/Namespace"
+    - $ref: "#/parameters/Name"
+    post:
+      summary: "Instantiate a BuildConfig object."
+      description: "Instantiate a BuildConfig object. This will result in new Build being created and started."
+      parameters:
+      - $ref: "#/parameters/BuildRequest"
+      responses:
+        200:
+          description: "OK"
+          schema:
+            type: object
+            properties:
+              BuildName:
+                type: string
+        401:
+          $ref: "#/responses/401"
+        500:
+          $ref: "#/responses/500"
+  /namespace/{namespace}/builds/{name}:
```

Comment:
/namespace/{namespace}/builds/{name}, instead of name should this be build_id or build_number